compiler/src/iree/compiler/Codegen/Common/VectorLayoutAnalysis.h

```diff
@@ -0,0 +1,138 @@
+///
+/// To start, the user must define some anchor points using `setAnchor`,
+/// anchoring a layout to a specific value. The analysis can then be ran using
+/// `run`, at which point it will use it's rules to infer other values in the
+/// program. Example:
+///
+///    %root = vector.transfer_read
+///      |
+///      --> anchored to layout L
+///    %root2 = vector.transfer_read
+///    %c = arith.mulf %root, %b
+///          |
+///          --> %root, %b and %c must have the same layout
+///    %e = arith.divf %b, %root2
+///          |
+///          --> %root2, %b and %e must have the same layout
+///
+/// Here, the user provided an anchor point for %root, fixing it's layout to L.
+/// The layout then uses it's inference rules to find the layout of other
+/// values:
+///
+///    %root = vector.transfer_read
+///     |
+///     --> infered to layout L
+///    %root2 = vector.transfer_read
+///     |
+///     --> infered to layout L
+///    %c = arith.mulf %root, %b
+///     |
+///     --> infered to layout L
+///    %e = arith.divf %b, %root2
+///     |
+///     --> infered to layout L
+///
+/// If at any point, a value has a layout, but the user of that value requires
+/// a different layout, the analysis inserts a resolution operation. This
+/// resolution operation is `iree_vector_ext.layout_conflict_resolution`.
+/// For Example:
+///
+/// %0 = vector.transfer_read
+///  |
+///  --> anchored to layout L
+/// %1 = vector.transfer_read
+///  |
+///  --> anchored to layout L'
+///  arith.addf %0, %1
+///     |
+///     --> %0 and %1 must have the same layout
+///
+/// To resolve the conflict, the analysis chooses one of the layouts, say
+/// L, and inserts a resolution operation to convert the other layout to L.
+///
+/// %0 = vector.transfer_read
+///  |
+///  --> anchored to layout L
+/// %1 = vector.transfer_read
+///  |
+///  --> anchored to layout L'
+/// %resolved = iree_vector_ext.layout_conflict_resolution %1
+///  |
+///  --> infered to layout L
+/// arith.addf %0, %resolved
+///
+/// The analysis itself will not try to resolve the conflict, but instead
+/// leaves it to the user to resolve the conflict.
+class VectorLayoutAnalysis {
+public:
+  VectorLayoutAnalysis(Operation *root) : root(root) {}
+
+  /// Fix the layout for a specific value. The layout must implement
+  /// VectorLayoutInterface.
+  template <typename T>
+  void setAnchor(Value val, T layout) {
+    assert(isa<VectorLayoutInterface>(layout) &&
+           "expected layout to implement VectorLayoutInterface");
+    anchors[val] = cast<VectorLayoutInterface>(layout);
+  }
+
+  /// Run the analysis. The analysis expects that the user has set some anchor
+  /// points and is trying to infer the layout of other values.
+  LogicalResult run();
+
+  /// Get the infered layout of a specific value. This should only be called
+  /// after the analysis has been run.
+  template <typename T>
+  T getLayout(Value val) {
```

Comment:
can this return an std::optional instead?

Reply:
The layout is expected to be an Attribute. Attributes already have a null check implemented so I'm not sure if we need that.